src/lib/optimizer/strategy/duplicate_elimination_rule.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "abstract_rule.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+
+/**
+ * This optimizer rule eliminates duplicates of the LQP by using sub-plan memoization:
+ * This process uses depth-first traversal and starts integrating sub-plans into the
+ * memoization structure at the leaves of the overall LQP and works upwards to the root.
+ */
+
+class DuplicateEliminationRule : public AbstractRule {
```

Comment:
Naming: "DuplicateEliminationRule" immediately made me think of a rule that eliminates duplicate Projections. SubqueryRedundancyEliminationRule or something?